pincer/utils/event_mgr.py

```diff
@@ -0,0 +1,211 @@
+# Copyright Pincer 2021-Present
+# Full MIT License can be found in `LICENSE` at the project root.
+
+from asyncio import Event, wait_for as _wait_for, get_running_loop
+from typing import Any, Callable, Union
+
+
+class _DiscordEvent(Event):
+    """
+    Attributes
+    ----------
+    return_value : Optional[str]
+        Used to store the arguments from ``can_be_set`` so they can be
+        returned later.
+    """
+
+    def __init__(
+        self,
+        event_name: str,
+        check: Union[Callable[[Any], bool], None]
+    ):
+        """
+        Parameters
+        ----------
+        event_name : str
+            The name of the event.
+        check : Optional[Callable[[Any], bool]]
+            ``can_be_set`` only returns true if this function returns true.
+            Will be ignored if set to None.
+        """
+        self.event_name = event_name
+        self.check = check
+        self.return_value = None
+        super().__init__()
+
+    def can_be_set(self, event_name: str, *args) -> bool:
+        """
+        Parameters
+        ----------
+        event_name : str
+            The name of the event.
+        *args : Any
+            Arguments to evaluate check with.
+
+        Returns
+        -------
+        bool
+            Whether the event can be set
+        """
+        if self.event_name != event_name:
+            return False
+
+        self.return_value = args
+
+        if self.check:
+            return self.check(*args)
+
+        return True
+
+
+class EventMgr:
+    """
+    Attributes
+    ----------
+    stack : List[_DiscordEvent]
+        The List of events that need to be processed.
+    """
+
+    def __init__(self):
+        self.stack = []
```

Comment:
@Arthurdw Using custom stack class maybe? not sure

Reply:
The class uses methods from the list. (`list.remove()`). I will rename the variable so it makes more sense.